blog: blogpost summarizing achievements of revival milestone 3 (#246)
MovingBlocks/movingblocks.github.com: blog/2024-06-26-third-revival-milestone-complete/index.md

---
posttype: blog
title: "Exploring Revival Milestone 3 - Upgrades, Fixes and Improvements Left, Right and Center"
cover: "./cover.jpg"

> description: "Exploring Terasology Revival Milestone 3: Our continued focus on performance and quality was supported by various fixes left and right. We’re celebrating our achievements and planning our path ahead. We’re nearing the gameplay content work phase: Almost Ready!"

Reviewer: > We’re nearing the gameplay content work phase

This is technically accurate but maybe a bit too much "project manager style" for an outreach blog post. I am not sure about the best way to word this but I think we should avoid the term "work phase" where possible. It implies two things that I'm not sure we intend:
1. That we see this as "work". Well, it is work, in the literal sense. One of the things I've found common with discussing hobby-based projects is that you care more about the goal than the means. It is "work" to get there but what we're doing is building a game, not generic "work" as-if it's just any old job. Maybe it's just for me that "work" has connotations of the standard 9-5 day job though.
2. The work "phase" has a defined end point. We have aspirations (that I hope we can meet) but are not committed to a fixed time span.

Maybe something like the following might be better?
```suggestion
description: "Exploring Terasology Revival Milestone 3: Our continued focus on performance and quality was supported by various fixes left and right. We’re celebrating our achievements and planning our path ahead. We’re laying the foundations for new and improved gameplay: Almost Ready!"
```

Author: What I was going for was rather "content work" than "work phase", but I wasn't happy with that wording either. I like "laying foundations" 🙂

author: "Niruandaleth"
date: "2024-06-26"
tags: ["Announcement"]
---

_Exploring Terasology Revival Milestone 3: Our continued focus on performance and quality was supported by various fixes left and right.
We’re celebrating our achievements and planning our path ahead.

> We’re nearing the gameplay content work phase: Almost Ready!_

Reviewer: See my comment above about the wording of this particular line.

Now we are excited to announce the completion of our third revival milestone.
This phase continued our focus on performance and quality while new contributors joined in and provided additional bug fixes and refactoring progress.
Right now, we're gearing up for the fourth milestone and are excited about the journey ahead paving the way for innovative content and features 🤓
The journey so far has been both challenging and rewarding and we are grateful for the continued support and contributions from our amazing community. 
Together, we can make Terasology better than ever! 🙌

## 📜 Definition of Done: A Checklist of Progress

🎉 In total, we merged [27 engine PRs](https://github.com/MovingBlocks/Terasology/pulls?page=1&q=is%3Apr+sort%3Aupdated-desc+is%3Amerged+merged%3A2023-11-20..2024-06-16), [78 module PRs](https://github.com/search?q=+org%3Aterasology+is%3Apr+merged%3A2023-11-20..2024-06-16+&type=pullrequests&p=1), and [20 PRs](https://github.com/search?q=-repo%3Amovingblocks%2Fterasology+org%3Amovingblocks+is%3Apr+merged%3A2023-11-20..2024-06-16+&type=pullrequests&p=3) on libraries and side projects such as gestalt, the launcher, or the module site.

> After the migration to Java 17 as part of our last milestone, in this milestone we spent time resolving issues arising from this migration 🐛

Reviewer: ```suggestion
After the migration to Java 17 as part of our last milestone, in this milestone we spent time resolving issues arising from the migration 🐛
```

The double-use of "this" makes for a confusing context while reading. I prefer "the migration" in this case.

If you want to contribute to our efforts of improving performance, come talk to us on our [Discord performance channel](https://discord.com/channels/270264625419911192/882575244215341097) or check out [Terasology#5150](https://github.com/MovingBlocks/Terasology/issues/5150) which includes todos and various spots to understand, document, refactor, and improve.

In addition, the user experience got improved for both players and developers.
Players can now enjoy a more streamlined advanced game setup with integrated configuration and preview options for the world they are about to play in. Feel free to check it out and tell us what you think! 🎮
Developers on the other hand are now enabled to get their PRs in much quicker after the flaky engine integration tests were split out into an optional CI step. Furthermore, formerly broken Jenkins pipelines are now up and running again and await your PRs 💻

Thanks to various new contributors as well as our core revival team, we merged many many fixes this milestone 💪
They removed client and server bugs, in-game content and graphics bugs as well as vulnerable web dependencies and broken documentation links 🐞

> Thank you to _@Khaled-Dridi_ and _@Imitater967_ for their fixes for the windows fullscren bug and no longer working visual effects and command line arguments 👏

Reviewer: ```suggestion
Thank you to _@Khaled-Dridi_ and _@Imitater967_ for their fixes for the fullscreen bug on Windows and no longer working visual effects and command line arguments 👏
```

Windowed Fullscreen (a borderless window occupying the entire screen) and fullscreen on Windows are two different things. Maybe this would make it clearer that the fullscreen bug was Windows-only?

Author: absolutely! I think that was one of the last things I added in and my brain was a bit blank by then 😅 good catch!

Special thanks go to _@priya4991_, the author of the partial fix for M1 support 💚 This problem is not fully gone yet, so if you develop on Mac and want to contribute to Terasology, any help with getting Terasology on M1 Macs to build and run again would be highly appreciated! 🍎

Next to these bigger efforts, a lot of smaller discussions and activities happened across our GitHub repositories and Discord channels.
_@BSA_ and _@Niruandaleth_ met once per week virtually to discuss milestone progress, agree on strategic questions, and collaborate on blockers and reviews 📅
Over the course of our third revival milestone, _@Cervator_ provided guidance for new community members and revivers alike, sharing his knowledge of the project history and in particular its infrastructure ⌨

> Revival milestone contributor already for the previous milestone, _@soloturn_, considerably contributed to this milestone as well by updating dependencies, facilitating migrations to Kotlin, and streamlining PR workflows by addressing issues flagged by PMD 🤓

Reviewer: ```suggestion
Returning from our previous milestone, _@soloturn_ again contributed considerably to this milestone by updating dependencies, facilitating migrations to Kotlin and streamlining PR workflows by addressing issues flagged by PMD 🤓
```

I prefer this wording personally but it essentially means the same thing.

Author: Again a phrase I wasn't happy with, so thanks for improving it 👍

Recently joined contributors _@spookynutz_ and _@manumafe98_ authored various moduleland PRs refactoring components and test annotations, thus achieving a more consistent use of ECS and our module testing framework 💪

> Long-time contributor _@engiValk_ visited the project to fix the water super jump bug reported by @SomethingSomethingV2, improving gameplay consistency 🔨

Reviewer: Here it seems that `@SomethingSomethingV2` is not italicised, whilst the other mentioned names are. Should all contributor names be formatted the same, or are there particular rules as to when to do this?

Author: No, you're absolutely right, it should be italicised like the others, I just forgot to do that in that spot.

_@qwc_ will host the Valheim server again for our third revival milestone completion celebrations, fostering team spirit and camaraderie 🕹

In conclusion, Milestone 3 represents a testament to our commitment to excellence in performance, quality, and user experience.
Through collaborative efforts and innovative solutions and with the help of warmly-welcomed new contributors, we continue to push our Terasology revival further 🌈

## 🚀 Terasology Revival Milestone 4: Moving Towards Gameplay Content Work

As we stride into Milestone 4 of our Terasology revival journey, our focus sharpens on preparing the groundwork for working again on gameplay content.
Again, our repositories are wide open for new contributions by friendly people joining our gooey community! 🌟 Whether you're a seasoned developer or just getting started, there's a place for you in our dynamic community.
Here’s a detailed breakdown of our priorities and the contributors driving each task. If you'd like to contribute to a specific task, feel free to ping the respective revivers on our Discord.

### 🛠 Infrastructure & Stability Enhancements

> Led by @BenjaminAmos (BSA), we will be migrating to gestalt 8. This involves upgrading dependencies to gestalt 8.0.0-SNAPSHOT and transitioning from nui-gestalt7 to nui-gestalt.

Reviewer: ```suggestion
Led by @BenjaminAmos (BSA), we will be migrating to the latest iteration of our gestalt library (gestalt 8). This involves upgrading dependencies to gestalt 8.0.0-SNAPSHOT and transitioning from nui-gestalt7 to nui-gestalt.
```

I would recommend a bit more context about gestalt, although that's a bit tricky since, as it's name quite accurately describes, it is a collection of bits which make more sense when combined together. As a minimum, I think it's worth mentioning that this is an iterative improvement on our self-built library, rather than some fancy new miracle framework.

I am not too sure that the fine technical details of this migration belong in a blog post. It would be better if we had a GitHub issue tracking the effort that you could then to link to in the blog post if people are interested in more information.

Author: Totally fair to add more context. With the last bit removed based on your comment below, I think the level of technical details is acceptable.

Annotating all relevant annotations and base classes with @Index and @IndexInherited respectively ensures efficient class discovery.

> Additionally, replacing deprecated API references with org.terasology.context.annotation.API standardizes our codebase for future enhancements. 🛠️

Reviewer: Technically we never deprecated the old `@API` annotation. Gestalt 8 just moves it into a different package.

Author: removed 👍

@jdrueckert (Niruandaleth) will address the flaky integration tests that have been bothering for a while now, particularly those causing sporadic NPEs in chunk lighting.
This investigation ensures our testing suite remains reliable, providing a stable foundation for ongoing development efforts. 🔍

In preparation for future releases, @skaldarnar is enhancing our launcher to support multiple Java versions seamlessly.
This feature allows to start old and new Terasology releases from the same launcher without the need to have a specific Java version installed. 🚀

### 🔍 Performance Optimization & Core Improvements

@BenjaminAmos (BSA) and @jdrueckert (Niruandaleth) continue to investigate performance bottlenecks, particularly focusing on chunk generation and loading stuttering identified in CoreGameplay.
This initiative aims to deliver smoother gameplay experiences across various hardware configurations.

> Planned efforts include refactoring code in the top ten critical performance areas as documented in [Terasology#5150]() for optimization potential.

Reviewer: ```suggestion
Planned efforts include refactoring code in the top ten critical performance areas as documented in [Terasology#5150](https://github.com/MovingBlocks/Terasology/issues/5150) for optimization potential.
```
I think you forgot to include the link here.

Author: indeed I did

Cleaning up the codebase further and improving sprint/crouch mechanics are additional key goals set by @jdrueckert (Niruandaleth).
This includes removing obsolete controller settings and graphic configurations (#5152), streamlining our codebase for improved maintainability and performance 🧹
Movement enhancements put an emphasis on elevating user interaction and paving the way for future gameplay enhancements 🏃‍♂️ 

Finally, @BenjaminAmos (BSA) will begin researching requirements for the clean separation of server and client to simplify networking in multiplayer setups.

### ⚙ Quality Assurance and Engine Refactorings

> Like before, if time permits, @jdrueckert (Niruandaleth) aims to continue with quality improvements and code refactorings.

Reviewer: ```suggestion
As before, @jdrueckert (Niruandaleth) aims to continue with quality improvements and code refactorings.
```

We are not constraining this to a fixed timescale, so time permitting is a slightly misleading thing to say here. What I understand this to mean is that this is not a requirement for finishing our milestone. Using some tentative language may be enough to bring that across more clearly but I am not sure on this one.

Author: I couldn't really think of how to phased that more tentatively... Skal argued that "to aim for something" already is relatively "weak", though, for instance compared to "to strive for something"
